lambda/translmod.ml

```diff
@@ -1660,24 +1672,43 @@ let print_cycle ppf cycle =
     (pp_print_list ~pp_sep print_ident) cycle
     pp_sep ()
     (Ident.name @@ fst @@ List.hd cycle)
-(* we repeat the first element to make the cycle more apparent *)
+
+let rec collect_components = function
+  | Pident id -> [Ident.name id]
+  | Pdot (p, s) -> collect_components p @ [s]
+  | Papply (p, _) -> collect_components p
+  | Pextra_ty (p, _) -> collect_components p
+
+let get_relative_path top_module path =
+  let comps = collect_components path in
+  let comps =
+    match comps with
+    | h :: t when h = top_module -> t
+    | _ -> comps
+  in
+  String.concat "." comps
+
 
 let explanation_submsg (id, unsafe_info) =
   match unsafe_info with
-  | Unnamed -> assert false (* can't be part of a cycle. *)
-  | Unsafe {reason;loc;subid} ->
-      let print fmt =
-        let printer = doc_printf fmt
-            Style.inline_code (Ident.name id)
-            Style.inline_code (Ident.name subid) in
-        Location.mkloc printer loc in
-      match reason with
-      | Unsafe_module_binding ->
-          print "Module %a defines an unsafe module, %a ."
-      | Unsafe_functor -> print "Module %a defines an unsafe functor, %a ."
-      | Unsafe_typext ->
-          print "Module %a defines an unsafe extension constructor, %a ."
-      | Unsafe_non_function -> print "Module %a defines an unsafe value, %a ."
+  | Unnamed ->
+      assert false
+  | Unsafe {reason; loc; path} ->
+      let top_level = Ident.name id in
+      let name_to_print = get_relative_path top_level path in
+      let reason = match reason with
+        | Unsafe_module_binding -> "defines an unsafe module"
+        | Unsafe_functor -> "defines an unsafe functor"
+        | Unsafe_typext -> "defines an unsafe extension constructor"
+        | Unsafe_non_function -> "defines an unsafe value"
```

Comment:
This refactoring is not a good idea, because @Octachron wants to keep full grammatical sentences in error-message string literals, to be able to possibly do internationalization. You should revert to the previous shape of the code, just with an extra call to `get_relative_path`.

Reply:
Internationalizations is a side bonus, my main rational is that keeping whole proposition makes it possible to do copy-editing on the error messages without having to check all execution paths.